src/uucore/src/lib/lib.rs

```diff
@@ -67,10 +67,63 @@ pub use crate::features::wide;
 
 use std::ffi::OsString;
 
+pub enum InvalidEncodingHandling {
+    Ignore,
+    ConvertLossy,
+    Panic,
+}
+
+#[must_use]
+pub enum ConversionResult {
+    FullyDecoded(Vec<String>),
+    Lossy(Vec<String>),
+}
+
+impl ConversionResult {
+    pub fn accept_any(&self) -> &Vec<String> {
+        match self {
+            Self::FullyDecoded(result) => result,
+            Self::Lossy(result) => result,
+        }
+    }
+}
+
 pub trait Args: Iterator<Item = OsString> + Sized {
-    fn collect_str(self) -> Vec<String> {
-        // FIXME: avoid unwrap()
-        self.map(|s| s.into_string().unwrap()).collect()
+    /// Converts each iterator to a String and collects these into a vector
+    /// On failure, the result will may be lossy or incomplete
+    /// # Arguments
+    /// * `handling` - This switch allows to define the behavior, when invalid encoding is encountered
+    fn collect_str(self, handling: InvalidEncodingHandling) -> ConversionResult {
+        let mut full_conversion = true;
+        let result_vector: Vec<String> = self
+            .map(|s| match s.into_string() {
+                Ok(string) => string,
+                Err(s_ret) => {
+                    full_conversion = false;
+                    let lossy_conversion = s_ret.to_string_lossy();
+                    eprintln!(
+                        "Input with broken encoding occured! (s = '{}') ",
+                        &lossy_conversion
+                    );
+                    match handling {
+                        InvalidEncodingHandling::Ignore => String::new(),
+                        InvalidEncodingHandling::ConvertLossy => lossy_conversion.to_string(),
+                        InvalidEncodingHandling::Panic => {
+                            panic!("Broken encoding found but caller cannot handle it")
+                        }
+                    }
+                }
+            })
+            .collect();
+        match full_conversion {
+            true => ConversionResult::FullyDecoded(result_vector),
+            false => ConversionResult::Lossy(result_vector),
+        }
+    }
+
+    /// convience function for a more slim interface
+    fn collect_str_lossy(self) -> ConversionResult {
+        collect_str(InvalidEncodingHandling::ConvertLossy)
     }
```

Comment:
@sylvestre what do you think of this? It allows a slim interface with convenience functions like collect_str_lossy while still giving the callers control of what should happen in cases of error. Also, as a caller, you will get a result-like type which allows you to react. There are two example-usages: head and dirname. If you like the design, i will finish this by adding testing and changing all callers to the new API (i would figure the desired behaviour of each command to match the original coreutils).